File: Problems/0039.Combination-Sum/CombinationSum.py
from typing import List, Set

class Solution:

    # Runtime: 56 ms, faster than 87.55% of Python3 online submissions for Combination Sum.
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        print("XXX ",candidates, target )
        res = []
        for i, n in enumerate(candidates):
            if n == target:
                res.append([n])
            elif n < target:
                subRes = self.combinationSum(candidates[i:], target - n)
                res.extend([[n] + r for r in subRes])


        print("====", candidates, target, res)
        return res

    # Runtime: 56 ms, faster than 87.55% of Python3 online submissions for Combination Sum.
    def combinationSumV2(self, candidates: List[int], target: int) -> List[List[int]]:
        def combinHelper(candidates, target, prefixRes):
            for i, n in enumerate(candidates):
                if n == target:
                    res.append(prefixRes + [n])
                elif n < target:
                    combinHelper(candidates[i:], target - n, prefixRes + [n])
        res = []
        combinHelper(candidates, target, [])
        return res

    # Runtime: 48 ms, faster than 96.82% of Python3 online submissions for Combination Sum.
    def combinationSumV3(self, candidates: List[int], target: int) -> List[List[int]]:
        def combinHelper(candidates, target, prefixRes, res):
            for i, n in enumerate(candidates):
                if n == target:
                    res.append(prefixRes + [n])
                elif n < target:
                    combinHelper(candidates[i:], target - n, prefixRes + [n], res)
        res = []
        combinHelper(candidates, target, [], res)
        return res

File: leetcode/0039.Combination-Sum/CombinationSum.go
package leetcode

// Runtime: 4 ms, faster than 79.63% of Go online submissions for Combination Sum.
func combinationSum(candidates []int, target int) [][]int {
	res := [][]int{}
	for i, n := range candidates {
		if n == target {
			res = append(res, []int{n})
		} else if n < target {
			subres := combinationSum(candidates[i:], target-n)
			for _, sub := range subres {
				res = append(res, append([]int{n}, sub...))
			}
		}
	}
	return res
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Combination Sum.
func combinHelper(candidates []int, target int, prefixRes []int, res *[][]int) {
	for i, n := range candidates {
		if n == target {
			copyPrefix := make([]int, len(prefixRes))
			copy(copyPrefix, prefixRes)
			*res = append(*res, append(copyPrefix, n))
		} else if n < target {
			combinHelper(candidates[i:], target-n, append(prefixRes, n), res)
		}
	}
}

func combinationSumV2(candidates []int, target int) [][]int {
	res := [][]int{}
	combinHelper(candidates, target, []int{}, &res)
	return res
}
